File: dsa/week(4)/Multipath_Inher.js
class Vehicle {
    #maxSpeed; 
    constructor(z) {
        console.log(`Vehicle's Parametrized constructor ${z}`);
        this.#maxSpeed = z;
    }
    print() {
        console.log("Vehicle");
    }
}
class Car extends Vehicle {
    constructor() {
        super(3); 
        this.numGears = 0;
        console.log("Car's default constructor");
    }
    print() {
        console.log(`car`);
    }
}

class Truck extends Vehicle {
    constructor() {
        super(4); 
        console.log("Truck's constructor");
    }
}

class Bus extends Car {
    constructor() {
        super(); // Calls Car's constructor

        Vehicle.call(this, 5); // Explicitly calls Vehicle's constructor
		
        console.log("Bus's constructor");
    }
}

// Create a Bus instance
const b = new Bus();
b.print();

File: dsa/week(4)/Multipath_Inher.py
class Vehicle:
    def __init__(self, z):
        self.__maxSpeed = z 
        self.numTyres = 0 
        self.color = ""  
        print(f"Vehicle's Parametrized constructor {z}")

    def print(self):
        print("Vehicle")


class Car(Vehicle):
    def __init__(self):
        super().__init__(3)  # Calls Vehicle's constructor with parameter
        self.numGears = 0
        print("Car's default constructor")

    def print(self):
        print(f"NumTyres : {self.numTyres}")
        print(f"Color : {self.color}")
        print(f"Num gears : {self.numGears}")


class Truck(Vehicle):
    def __init__(self):
        super().__init__(4)  # Calls Vehicle's constructor with parameter
        print("Truck's constructor")


class Bus(Car, Truck):
    def __init__(self):
        Car.__init__(self)  
        Truck.__init__(self)
        print("Bus's constructor")


# Create a Bus instance
b = Bus()
b.print()
